python/Ganga/test/GPI/FileTests/TestLocalFileClient.py

```diff
@@ -0,0 +1,102 @@
+from __future__ import absolute_import
+from Ganga.testlib.GangaUnitTest import GangaUnitTest
+from Ganga.testlib.file_utils import generate_unique_temp_file
+
+import datetime
+import time
+import os
+import shutil
+import copy
+
+class TestLocalFileClient(GangaUnitTest):
+    """test for sjid in filename names explain each test"""
+
+    _managed_files = []
+
+    # Num of sj in tests
+    sj_len = 3
+
+    # This sets up a LocalFileConfiguration which works by placing a file on local storage somewhere we can test using standard tools
+    LocalFileConfig = {'fileExtensions': [''],
+                       'uploadOptions': {},
+                       'backendPostprocess': {'LSF': 'client', 'LCG': 'client', 'ARC': 'client', 'Dirac': 'client',
+                                              'PBS': 'client', 'Interactive': 'client', 'Local': 'client', 'CREAM': 'client'}}
+
+    def setUp(self):
+        """
+        Configure the LocalFile for the test
+        """
+        extra_opts=[('PollThread', 'autostart', 'False'),
+                    ('Local', 'remove_workdir', 'False'),
+                    ('TestingFramework', 'AutoCleanup', 'False'),
+                    ('Output', 'LocalFile', self.LocalFileConfig),
+                    ('Output', 'FailJobIfNoOutputMatched', 'True')]
+        super(TestLocalFileClient, self).setUp(extra_opts=extra_opts)
+
+    @staticmethod
+    def cleanUp():
+        """ Cleanup the current temp objects """
+
+        from Ganga.GPI import jobs
+        for j in jobs:
+            shutil.rmtree(j.backend.workdir, ignore_errors=True)
+            j.remove()
+
+        for file_ in TestLocalFileClient._managed_files:
+            os.unlink(file_)
+        TestLocalFileClient._managed_files = []
+
+    def test_a_testClientSideSubmit(self):
+        """Test the client side code whilst stil using the Local backend"""
+
+        from Ganga.GPI import LocalFile, Job, ArgSplitter
+
+        TestLocalFileClient.cleanUp()
+
+        _ext = '.root'
+        file_1 = generate_unique_temp_file(_ext)
+        file_2 = generate_unique_temp_file(_ext)
+        TestLocalFileClient._managed_files.append(file_1)
+        TestLocalFileClient._managed_files.append(file_2)
+
+        j = Job()
+        j.inputfiles = [LocalFile(file_1), LocalFile(file_2)]
+        j.splitter = ArgSplitter(args = [[_] for _ in range(0, TestLocalFileClient.sj_len) ])
+        j.outputfiles = [LocalFile(namePattern='*'+_ext)]
+        j.submit()
+
+    def test_b_testClientSideComplete(self):
```

Comment:
Obviously it has the option 3 but you seem to be ignoring the difference between writing *all* tests in this form and reserving it for those cases where it is necessary. I have mentioned this in several PRs in the past; it's nothing new. The goalposts have not been moved, they've just been ignored. Regardless, my comments above on this code are not a diktat, they are advice on your code which you asked for by opening this PR.

For the record, here's the full `TestLocalFileClient.py` test converted to a pytest style:
```python
import os

import pytest

from Ganga.GPIDev.Base.Proxy import stripProxy
from Ganga.testlib.decorators import add_config
from Ganga.testlib.monitoring import run_until_completed


@pytest.mark.parametrize('post_process_location', ['client', 'WN'])
@add_config([('Output', 'FailJobIfNoOutputMatched', 'True')])
def testClientSide(gpi, post_process_location, tmpdir):
    from Ganga.GPI import LocalFile, Job, ArgSplitter, config
    config['Output']['LocalFile'] = {'fileExtensions': [''],
                                     'uploadOptions': {},
                                     'backendPostprocess': {'Local': post_process_location}}

    file_1 = tmpdir.join('file_1.root')
    file_1.write('blah blah')
    file_2 = tmpdir.join('file_2.root')
    file_2.write('blah blah')

    j = Job()
    j.inputfiles = [LocalFile(str(file_1)), LocalFile(str(file_2))]
    j.splitter = ArgSplitter(args=[[i] for i in range(3)])
    j.outputfiles = [LocalFile(namePattern='*.root')]
    j.submit()

    assert run_until_completed(j)

    for sj in j.subjobs:
        output_dir = stripProxy(sj).getOutputWorkspace(create=False).getPath()
        assert os.path.isdir(output_dir)

        # Check that the files were placed in the correct place on storage
        for file_ in j.inputfiles:
            assert os.path.isfile(os.path.join(output_dir, file_.namePattern))

        # Check that wildcard expansion happened correctly
        assert len(stripProxy(stripProxy(sj).outputfiles[0]).subfiles) == 2

        assert len(sj.outputfiles) == 2
```
It's perhaps not perfect since I whipped it up in 10 minutes or so but it's *half* the length of the version in this PR, is between 2 and 3 seconds faster and allows the simple rerunning of a failing test.

I'm not suggesting that you replace your test with this but it's worth looking at to see what's possible for future.